blinding.go

```diff
@@ -0,0 +1,169 @@
+	// if the UnBlindedHopInfo wrapper is used and represents the blinded
+	// pub key if the BlindedHopInfo wrapper is used.
+	NodePub *btcec.PublicKey
+
+	// Payload is the data to be transported to the hop. It is the cleartext
+	// if the UnBlindedHopInfo wrapper is used and is the ciphertext if the
+	// BlindedHopInfo wrapper is used.
+	Payload []byte
+}
+
+// BlindedHopInfo represents a blinded node pub key along with the blinded data
+// for that node.
+type BlindedHopInfo hopInfo
+
+// UnBlindedHopInfo represents a real node pub key along with the plaintext data
+// for that node.
+type UnBlindedHopInfo hopInfo
+
+// BuildBlindedPath creates a new BlindedPath from a list of UnBlindedHopInfo
+// and a session key.
+func BuildBlindedPath(sessionKey *btcec.PrivateKey,
+	paymentPath []*UnBlindedHopInfo) (*BlindedPath, error) {
+
+	if len(paymentPath) < 1 {
+		return nil, fmt.Errorf("at least 1 hop is required to create " +
+			"a blinded path")
+	}
+
+	bp := BlindedPath{
+		IntroductionPoint: paymentPath[0].NodePub,
+		BlindingPoint:     sessionKey.PubKey(),
+		BlindedHops:       make([]*BlindedHopInfo, len(paymentPath)),
+	}
+
+	keys := make([]*btcec.PublicKey, len(paymentPath))
+	for i, p := range paymentPath {
+		keys[i] = p.NodePub
+	}
+
+	hopSharedSecrets, err := generateSharedSecrets(keys, sessionKey)
+	if err != nil {
+		return nil, fmt.Errorf("error generating shared secret: %v",
+			err)
+	}
+
+	for i, hop := range paymentPath {
+		blindedInfo, err := BlindHopInfo(hop, hopSharedSecrets[i])
+		if err != nil {
+			return nil, err
+		}
+
+		bp.BlindedHops[i] = blindedInfo
+	}
+
+	return &bp, nil
+}
+
+// BlindHopInfo uses the given sharedSecret to blind the UnBlindedHopInfo and
+// returns the associated BlindedHopInfo.
+func BlindHopInfo(info *UnBlindedHopInfo, sharedSecret Hash256) (
+	*BlindedHopInfo, error) {
+
+	blindedData, err := BlindData(sharedSecret, info.Payload)
+	if err != nil {
+		return nil, err
+	}
+
+	blindedNodeKey := BlindNodeID(sharedSecret, info.NodePub)
+
+	return &BlindedHopInfo{
+		NodePub: blindedNodeKey,
+		Payload: blindedData,
+	}, nil
+}
+
+// BlindNodeID blinds the given public key using the provided shared secret.
+func BlindNodeID(sharedSecret Hash256,
+	pubKey *btcec.PublicKey) *btcec.PublicKey {
+
+	blindingFactorBytes := generateKey(routeBlindingHMACKey, &sharedSecret)
+
+	var blindingFactor btcec.ModNScalar
+	blindingFactor.SetBytes(&blindingFactorBytes)
+
+	return blindGroupElement(pubKey, blindingFactor)
+}
+
+// BlindData blinds/encrypts the given plain text data using the provided
+// shared secret.
+func BlindData(sharedSecret Hash256, plainTxt []byte) ([]byte, error) {
+	rho := generateKey("rho", &sharedSecret)
+	enc, err := chacha20polyEncrypt(rho[:], plainTxt)
+	if err != nil {
+		return nil, err
+	}
+
+	return enc, nil
+}
+
+// DecryptBlindedData decrypts the data encrypted by the creator of the blinded
+// route.
+func DecryptBlindedData(privKey SingleKeyECDH, ephemPub *btcec.PublicKey,
```

Comment:
Can we make this an exported function on `Router` so that we don't need to pass in the `privKey` here? For blinded forwarding in LND, we're going to need the [OnionProcessor](https://github.com/lightningnetwork/lnd/blob/262591c3331c2170697fe89b8c49ab857db0adf7/htlcswitch/hop/iterator.go#L130) to be able to decrypt these payloads, and it already has access to the sphinx router. 

While we _could_ thread through the [nodeKeyECDH](https://github.com/lightningnetwork/lnd/blob/master/server.go#L510) instance that we pass to the router to the onion processor [directly](https://github.com/lightningnetwork/lnd/blob/master/server.go#L581) (or export it in the `sphinx.Router`), this seems like a pretty logical method to add to the router?

Reply:
Suspect the same _may_ be true of encrypt fn but haven't gotten there yet :')

Reply:
cool yeah good idea - I added another commit that adds a method to Router that just calls this method.

Reply:
yergh sorry forgot to ask, can we do the same for `NextEphemeral` same case for using in lnd 🙏

Reply:
added! sorry, I should have known to add that too

Reply:
Thanks! Don't have perms to resolve this comment but all sorted :)